flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterExecutionDeploymentReconciliationTest.java

```diff
@@ -0,0 +1,186 @@
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests for the execution deployment-reconciliation logic in the {@link JobMaster}.
+ */
+public class JobMasterExecutionDeploymentReconciliationTest extends TestLogger {
+
+	private static final Time testingTimeout = Time.seconds(10L);
+
+	private static TestingRpcService rpcService;
+
+	private final HeartbeatServices heartbeatServices = new HeartbeatServices(Integer.MAX_VALUE, Integer.MAX_VALUE);
+
+	private final TestingHighAvailabilityServices haServices = new TestingHighAvailabilityServices();
+	private final SettableLeaderRetrievalService resourceManagerLeaderRetriever = new SettableLeaderRetrievalService();
+	private final TestingLeaderElectionService resourceManagerLeaderElectionService = new TestingLeaderElectionService();
+
+	@Rule
+	public final TestingFatalErrorHandlerResource testingFatalErrorHandlerResource = new TestingFatalErrorHandlerResource();
+
+	@Before
+	public void setup() {
+		haServices.setResourceManagerLeaderRetriever(resourceManagerLeaderRetriever);
+		haServices.setResourceManagerLeaderElectionService(resourceManagerLeaderElectionService);
+		haServices.setCheckpointRecoveryFactory(new StandaloneCheckpointRecoveryFactory());
+	}
+
+	@After
+	public void shutdown() {
+		rpcService.clearGateways();
+	}
+
+	@BeforeClass
+	public static void setupClass() {
+		rpcService = new TestingRpcService();
+	}
+
+	@AfterClass
+	public static void shutdownClass() throws ExecutionException, InterruptedException {
+		rpcService.stopService().get();
+	}
+
+	@Test
+	public void testExecutionDeploymentReconciliation() throws Exception {
+		JobMasterBuilder.TestingOnCompletionActions onCompletionActions = new JobMasterBuilder.TestingOnCompletionActions();
+		JobMaster jobMaster = createAndStartJobMaster(onCompletionActions);
+		JobMasterGateway jobMasterGateway = jobMaster.getSelfGateway(JobMasterGateway.class);
+		rpcService.registerGateway(jobMasterGateway.getAddress(), jobMasterGateway);
+
+		final JobMasterPartitionReleaseTest.AllocationIdsResourceManagerGateway resourceManagerGateway = createResourceManagerGateway();
+
+		final CompletableFuture<Void> taskSubmissionFuture = new CompletableFuture<>();
+		final CompletableFuture<ExecutionAttemptID> taskCancellationFuture = new CompletableFuture<>();
+		TaskExecutorGateway taskExecutorGateway = createTaskExecutorGateway(taskSubmissionFuture, taskCancellationFuture);
+		LocalUnresolvedTaskManagerLocation localUnresolvedTaskManagerLocation = new LocalUnresolvedTaskManagerLocation();
+
+		registerTaskExecutorAndOfferSlots(resourceManagerGateway, jobMasterGateway, taskExecutorGateway, localUnresolvedTaskManagerLocation);
+
+		taskSubmissionFuture.get();
```

Comment:
I think this test is unstable because the completion of this future does not guarantee that we are tracking the respective task. If you add a `Thread.sleep()` in line 161, then you can reproduce it. The reason is that we use a `whenCompleteAsync` when deploying an `Execution`.